File: src/main/java/com/integreety/yatspec/e2e/config/RabbitTemplateInterceptorConfig.java
package com.integreety.yatspec.e2e.config;

import com.integreety.yatspec.e2e.captor.rabbit.RabbitCaptor;
import com.integreety.yatspec.e2e.captor.rabbit.mapper.ExchangeNameDeriver;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.amqp.core.MessageBuilder;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.context.annotation.Configuration;

import javax.annotation.PostConstruct;
import java.util.List;

import static com.integreety.yatspec.e2e.captor.repository.model.Type.CONSUME;
import static com.integreety.yatspec.e2e.captor.repository.model.Type.PUBLISH;

@Slf4j
@Configuration
@ConditionalOnProperty(name = {"lsd.db.connectionstring"})
@ConditionalOnClass(RabbitTemplate.class)
@RequiredArgsConstructor
public class RabbitTemplateInterceptorConfig {

    private final List<RabbitTemplate> rabbitTemplates;
    private final RabbitCaptor rabbitCaptor;
    private final ExchangeNameDeriver exchangeNameDeriver;

    @PostConstruct
    public void configureRabbitTemplatePublishInterceptor() {
        rabbitTemplates.forEach(rabbitTemplate -> {
            rabbitTemplate.addBeforePublishPostProcessors(message -> {
                log.info("Rabbit message properties before publishing:{}", message.getMessageProperties());
                try {
                    final String exchangeName = exchangeNameDeriver.derive(message.getMessageProperties(), rabbitTemplate.getExchange());
                    rabbitCaptor.captureInteraction(exchangeName, MessageBuilder.fromMessage(message).build(), PUBLISH);
                } catch (final Throwable t) {
                    log.error(t.getMessage(), t);
                }
                return message;
            });
            rabbitTemplate.addAfterReceivePostProcessors(message -> {
                log.info("Rabbit message properties after receiving:{}", message.getMessageProperties());
                try {
                    final String exchangeName = exchangeNameDeriver.derive(message.getMessageProperties(), message.getMessageProperties().getReceivedExchange());
                    rabbitCaptor.captureInteraction(exchangeName, MessageBuilder.fromMessage(message).build(), CONSUME);
                } catch (final Throwable t) {
                    log.error(t.getMessage(), t);
                }
                return message;
            });
        });
    }
}

File: src/main/kotlin/io/lsdconsulting/lsd/distributed/interceptor/config/RabbitTemplateInterceptorConfig.kt
package io.lsdconsulting.lsd.distributed.interceptor.config

import io.lsdconsulting.lsd.distributed.access.model.InteractionType
import io.lsdconsulting.lsd.distributed.interceptor.captor.rabbit.RabbitCaptor
import io.lsdconsulting.lsd.distributed.interceptor.captor.rabbit.mapper.ExchangeNameDeriver
import org.springframework.amqp.core.Message
import org.springframework.amqp.core.MessageBuilder
import org.springframework.amqp.core.MessagePostProcessor
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.context.annotation.Configuration
import java.util.function.Consumer
import javax.annotation.PostConstruct

@Configuration
@ConditionalOnProperty(name = ["lsd.dist.connectionString"])
@ConditionalOnClass(RabbitTemplate::class)
open class RabbitTemplateInterceptorConfig(
    private val rabbitTemplates: List<RabbitTemplate>,
    private val rabbitCaptor: RabbitCaptor,
    private val exchangeNameDeriver: ExchangeNameDeriver,
) {
    @PostConstruct
    fun configureRabbitTemplatePublishInterceptor() {
        rabbitTemplates.forEach(Consumer<RabbitTemplate> { rabbitTemplate: RabbitTemplate ->
            rabbitTemplate.addBeforePublishPostProcessors(MessagePostProcessor { message: Message ->
                log().info(
                    "Rabbit message properties before publishing:{}",
                    message.messageProperties
                )
                try {
                    val exchangeName = exchangeNameDeriver.derive(message.messageProperties, rabbitTemplate.exchange)
                    rabbitCaptor.captureInteraction(
                        exchangeName,
                        MessageBuilder.fromMessage(message).build(),
                        InteractionType.PUBLISH
                    )
                } catch (t: Throwable) {
                    log().error(t.message, t)
                }
                message
            })
            rabbitTemplate.addAfterReceivePostProcessors(MessagePostProcessor { message: Message ->
                log().info(
                    "Rabbit message properties after receiving:{}",
                    message.messageProperties
                )
                try {
                    val exchangeName = exchangeNameDeriver.derive(
                        message.messageProperties,
                        message.messageProperties.receivedExchange
                    )
                    rabbitCaptor.captureInteraction(
                        exchangeName,
                        MessageBuilder.fromMessage(message).build(),
                        InteractionType.CONSUME
                    )
                } catch (t: Throwable) {
                    log().error(t.message, t)
                }
                message
            })
        })
    }
}
